ethereum/spec/src/property-test/java/tech/pegasys/teku/spec/datastructures/util/Mutator.java

```diff
@@ -0,0 +1,131 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.spec.datastructures.util;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import org.apache.tuweni.bytes.Bytes;
+
+public class Mutator {
+  /** The types of mutations that can occur. */
+  private enum Mutation {
+    REPLACE,
+    ADD_BEFORE,
+    ADD_AFTER,
+    DELETE
+  }
+
+  /** Likelihood that a specific mutation will happen. */
+  private static final Map<Mutation, Double> MUTATION_PROBABILITIES =
+      Map.of(
+          Mutation.REPLACE, 0.01,
+          Mutation.ADD_BEFORE, 0.001,
+          Mutation.ADD_AFTER, 0.001,
+          Mutation.DELETE, 0.001);
+
+  /** The total probability that a mutation of any kind will happen. */
+  private static final double MUTATION_PROBABILITY =
+      MUTATION_PROBABILITIES.values().stream().reduce(0.0, Double::sum);
+
+  /** If a random number falls in this range, that mutation should happen. */
+  private static final Map<Mutation, Range> RANGE_MAP = new HashMap<>();
+
+  static {
+    // Make a map of non-overlapping ranges. The range reflects the probability of that mutation.
+    // When fuzzing, a random number will be chosen. To determine which mutation should occur, we
+    // check which of these ranges the random number falls into. For example, you could expect this
+    // map to look something like:
+    //
+    // {
+    //    REPLACE: Range(0, 0.01),
+    //    ADD_BEFORE: Range(0.01, 0.011),
+    //    ADD_AFTER: Range(0.011, 0.012),
+    //    DELETE: Range(0.012, 0.013)
+    // }
+    double start = 0.0;
+    for (Map.Entry<Mutation, Double> entry : MUTATION_PROBABILITIES.entrySet()) {
+      RANGE_MAP.put(entry.getKey(), new Range(start, start + entry.getValue()));
+      start += entry.getValue();
+    }
+  }
+
+  /**
+   * A helper function for generating random input data. Instead of blinding making data, start with
+   * a known valid data and slightly mutate that. By default, you can expect a ~1% mutation rate.
+   *
+   * @param input The bytes to mutate, should resemble valid data.
+   * @param seed The seed that should be used for randomness, for reproducibility.
+   * @return A mutated version of the input.
+   */
+  public static Bytes mutate(final Bytes input, final int seed) {
+    final Random random = new Random(seed);
+    final List<Bytes> sections = new ArrayList<>();
+
+    for (final byte b : input.toArray()) {
+      final double randomNumber = Math.abs(random.nextGaussian());
+      if (randomNumber <= MUTATION_PROBABILITY) {
+
+        Mutation op = null;
+        for (final Map.Entry<Mutation, Range> entry : RANGE_MAP.entrySet()) {
```

Comment:
You might use a `TreeMap<Double,Mutation>.floorEntry` and you don't need a `Range` but just the upper bound of the probability

Reply:
Good idea. I'll look into that!

Reply:
Really great suggestion. Thanks. I think this looks a lot better. I ended up using `TreeMap.ceilingEntry` instead of `floorEntry` because calling `floorEntry` with a number greater than the largest key would return the last entry. I think `ceilingEntry` what you meant to suggest anyway, right?